File: packages/jsii-reflect/lib/type-instance.ts
import jsii = require('jsii-spec');
import { TypeReference } from './type-ref';
import { TypeSystem } from './type-system';

export class TypeInstance {
  public readonly type: TypeReference;

  constructor(
    public readonly system: TypeSystem,
    private readonly spec?: jsii.TypeInstance<jsii.TypeReference>) {
    this.type = new TypeReference(this.system, spec && spec.type);
  }

  public toString(): string {
    let description = this.type.toString();
    if (this.optional && !this.type.isAny) {
      description = `Optional<${description}>`;
    }
    if (this.promise) {
      description = `Promise<${description}>`;
    }
    return description;
  }

  /**
   * Indicates if this value is optional.
   */
  public get optional(): boolean {
    return this.spec != null && !!this.spec.optional;
  }

  /**
   * Indicates if this type refers to a promise.
   */
  public get promise(): boolean {
    return this.spec != null && !!this.spec.promise;
  }
}

File: packages/jsii-dotnet-jsonmodel/src/Amazon.JSII.JsonModel/Spec/TypeInstance.cs
using Newtonsoft.Json;

namespace Amazon.JSII.JsonModel.Spec
{
    [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
    public class TypeInstance
    {
        public TypeInstance(
            TypeReference type,
            bool? isOptional = null,
            bool? isPromise = null
        )
        {
            Type = type;
            IsOptional = isOptional;
            IsPromise = isPromise;
        }

        [JsonProperty("type")]
        public TypeReference Type { get; }

        [JsonProperty("optional", NullValueHandling = NullValueHandling.Ignore)]
        public bool? IsOptional { get; }

        [JsonProperty("promise", NullValueHandling = NullValueHandling.Ignore)]
        public bool? IsPromise { get; }
    }
}
